File: java/stack/com.main.stack/Stack.java
package com.main.stack;

public class Stack implements IStack {

    public Node top; // for cycles

    public Stack(){
        this.top = null;
    }


    @Override
    public void push(float value) {
        Node newItem = new Node(value);

        if(top == null) {
            top = newItem;
            return;
        }

        newItem.prev = top;
        top = newItem;
    }


    @Override
    public float pop(){
        if(this.size() == 1){
            float tmp = top.value;
            top = null;
            return tmp;
        }

        float tmp = top.value;
        top = top.prev;

        Node.decreaseLength();

        return tmp;
    }


    @Override
    public float top() {
        return top.value;
    }

    @Override
    public void printing() {
        System.out.print("[");

        for(Node item = top; item != null; item = item.prev){
            // Countdown starts from the top (Top)
            System.out.print(item.value);
            if(item.prev != null) System.out.print(", ");
        }

        System.out.println("]");
    }

    @Override
    public int size() {
        return Node.count;
    }
}

File: cpp/stack.cpp
/* 

-------------------------- CUSTOM DATA STRUCTURE - Stack -----------------------------
__ Created by https://github.com/bravTm

*/

#include <iostream>

using namespace std;


class Node {
public:
    float value;
    Node * prev;

    Node(float value){
        this->value = value;
        this->prev = NULL;
    }
};


class Stack {
public:
    Node * peak;
    int length;

    Stack() {
        this->length = 0;
        this->peak = NULL;
    }


    void push(float value){
        Node * newItem = new Node(value);

        if(peak == NULL){
            peak = newItem;
            return;
        }

        this->length++;

        newItem->prev = peak;
        peak = newItem;
    }


    float pop(){
        if(peak == NULL) return -0.333 * 3;

        float tmp = peak->value;
        Node * t = peak;
        peak = peak->prev;

        delete t;

        this->length--;

        return tmp;
    }


    float top(){
        return this->peak->value;
    }


    void printing(){
        cout << "[";

        for(Node * item = this->peak; item != NULL; item = item->prev){
            cout << item->value;

            if(item->prev != NULL) cout << ", ";
        }

        cout << "]\n";
    }

};




int main(){

    Stack st;

    st.push(10);
    st.push(3.4);
    st.push(3.8);
    st.push(0.22); // peak

    st.printing(); // [0.22, 3.8, 3.4, 10]

    cout << st.pop() << endl; // 0.22

    st.printing(); // [3.8, 3.4, 10]

    cout << st.top() << endl; // 3.8


    return 0;
}
